### PR TITLE
Issue #13553: False positive in FallThroughCheck on last case

### PR DESCRIPTION
**Pull request by aayushRedHat**
Related: https://github.com/checkstyle/checkstyle/issues/13553

Diff Regression config: https://gist.githubusercontent.com/aayushRedHat/849313f94f6df07629e364070d5ed19a/raw/3fc325f6c384a7ca918dc4a5875fa26cc1444679/gistfile1.txt

Diff Regression projects: https://gist.githubusercontent.com/aayushRedHat/085b2f4c08f7e2859c92e0d2cb2144c9/raw/358a901648397fd9f66978d4ce51636694546177/gistfile1.txt

### CODE REVIEW

**Comment by AayushSaini101** [comment]
@romani @rnveach @nrmancuso please help regarding the coverage issues in this PR

**Comment by nrmancuso** [comment]
> @romani @rnveach @nrmancuso please help regarding the coverage issues in this PR

Please explain what the issue is, and what you have tried so far to solve it.

**Comment by aayushRedHat** [comment]
> > @romani @rnveach @nrmancuso please help regarding the coverage issues in this PR
> 
> Please explain what the issue is, and what you have tried so far to solve it.

I am facing coverage issue, I have already added many test cases as @romani suggested but still coverage same 0.96. I need help to make coverage limit green

**Comment by nrmancuso** [comment]
@AayushSaini101 what code is causing the issue? Why can’t we hit this code? Have you tried to remove this code and generate a regression report?

**Comment by AayushSaini101** [comment]
> @AayushSaini101 what code is causing the issue? Why can’t we hit this code? Have you tried to remove this code and generate a regression report?

@nrmancuso Yes this code is causing an issue. I  asked the same question in the channel, Does the report depends on the code quality if yes. How may i know in what extent i need to improve the particular code

**Comment by nrmancuso** [comment]
@AayushSaini101 please take some time and research code coverage and Jacoco to deeply understand where your issue lies.

**Comment by romani** [comment]
@aayushRedHat , please generate HTML coverage report and publish it on your GitHub.io. we will guide you how to read it. If you share report we can help you during non keyboard time , a way quicker response time.

On side note: we use to have some service that showed coverage report for pull requests, but we might lost it or I can not find it in CI list.

**Comment by romani** [comment]
image

image
yellow means , that code is always goes inside block so expression is IF is never "false" but code expect it to be false some times, so we need to make test Input that will make "false" and skip execution of nested block. 


from pitest:
https://github.com/checkstyle/checkstyle/actions/runs/6832633689/job/18583535138?pr=14016#step:6:581
```
Source File: "FallThroughCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.coding.FallThroughCheck"
Method: "hasReliefComment"
Line Contents: "if (firstChild != null) {"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 424

Source File: "FallThroughCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.coding.FallThroughCheck"
Method: "hasReliefComment"
Line Contents: "if (previous != null) {"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 417

Source File: "FallThroughCheck.java"
Class: "com.puppycrawl.tools.checkstyle.checks.coding.FallThroughCheck"
Method: "hasReliefComment"
Line Contents: "if (text != null) {"
Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
Description: "removed conditional - replaced equality check with true"
Line Number: 426
```

**Comment by AayushSaini101** [comment]
> image
> 
> image yellow means , that code is always goes inside block so expression is IF is never "false" but code expect it to be false some times, so we need to make test Input that will make "false" and skip execution of nested block.
> 
> from pitest: https://github.com/checkstyle/checkstyle/actions/runs/6832633689/job/18583535138?pr=14016#step:6:581
> 
> ```
> Source File: "FallThroughCheck.java"
> Class: "com.puppycrawl.tools.checkstyle.checks.coding.FallThroughCheck"
> Method: "hasReliefComment"
> Line Contents: "if (firstChild != null) {"
> Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
> Description: "removed conditional - replaced equality check with true"
> Line Number: 424
> 
> Source File: "FallThroughCheck.java"
> Class: "com.puppycrawl.tools.checkstyle.checks.coding.FallThroughCheck"
> Method: "hasReliefComment"
> Line Contents: "if (previous != null) {"
> Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
> Description: "removed conditional - replaced equality check with true"
> Line Number: 417
> 
> Source File: "FallThroughCheck.java"
> Class: "com.puppycrawl.tools.checkstyle.checks.coding.FallThroughCheck"
> Method: "hasReliefComment"
> Line Contents: "if (text != null) {"
> Mutator: "org.pitest.mutationtest.engine.gregor.mutators.RemoveConditionalMutator_EQUAL_IF"
> Description: "removed conditional - replaced equality check with true"
> Line Number: 426
> ```

Thank you so much @romani . I am on it

**Comment by aayushRedHat** [comment]
Thanks @romani

**Comment by romani** [comment]
@aayushRedHat , please fix sonar violation

**Comment by AayushSaini101** [comment]
> @aayushRedHat , please fix sonar violation

Tried to Fix

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java:32]: What is reason of such updates?
If no good reason let's not change, we usually not changing old test data .
But we can change violation comments.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:18]: Please do not change existing test input code.
Change violation comments only.
If not possible, please share problem

**Review by romani — APPROVED** [review]
Green CI is awesome. Items

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java]: @romani Done, and removed violation

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:18]: @romani Done

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:18]: Please restore indentation, revert all charges in the this lines

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:24]: Restore as it was before.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:36]: Restore

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java:109]: Restore as it was before

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (AayushSaini101) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:24]: @romani but this violation should be removed because of fall thru comment above

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:24]: Violation line, line 25 should be changed, all other lines should not

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (AayushSaini101) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java:109]: @romani what about these

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java:109]: Make it fallthru, as it was before.

Comment (AayushSaini101) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java:109]: > Make it fallthru, as it was before.

Violation also need to be removed?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:24]: Done @romani

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java]: Done @romani

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java]: violation i always representation of actual behavior.
if there is violation from checkstyle, then we should put `// violation`.
if no violation, we remove comment.

Again: Code(even target comments) stays the same and only "// violation " comments changing their place or added or removed.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:45]: restore line 45.
if you need violation on such code please add "same" code to new Input

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java:109]: restore line 109

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java:109]: Done @romani

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:45]: Done @romani

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:46]: @aayushRedHat, revert all changes in line 45 and 46.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java:108]: line 108 should not be changed.

**Review by romani — CHANGES_REQUESTED** [review]
item:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughLastLineCommentCheck.java]: Done @romani

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough8.java:46]: Done @romani

**Comment by romani** [comment]
@nrmancuso , ready for review

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Before we go into implementation, I need clarity on the issue vs. what's in this PR.

1. The title of the issue is `False positive in FallThroughCheck on last case`, but I don't see anything limiting the new logic to only the last case. Please correct me if this is wrong. Judging from the changes here, the issue should probably be called 'Allow fall though comments anywhere in switch blocks'.
2. We must generate regression reports for any functional changes to a check, please do this.
3. I would expect a significant documentation update, since we are fundamentally changing the behavior of this check.

**Comment by aayushRedHat** [comment]
> Before we go into implementation, I need clarity on the issue vs. what's in this PR.
> 
> 1. The title of the issue is `False positive in FallThroughCheck on last case`, but I don't see anything limiting the new logic to only the last case. Please correct me if this is wrong. Judging from the changes here, the issue should probably be called 'Allow fall though comments anywhere in switch blocks'.
> 2. We must generate regression reports for any code changes to a check, please do this.
> 3. I would expect a significant documentation update, since we are fundamentally changing the behavior of this check.

@nrmancuso Issue Description needs to be updated here is the more info for this issue: https://github.com/checkstyle/checkstyle/issues/13553#issuecomment-1752129573

**Comment by aayushRedHat** [comment]
> > Before we go into implementation, I need clarity on the issue vs. what's in this PR.
> > 
> > 1. The title of the issue is `False positive in FallThroughCheck on last case`, but I don't see anything limiting the new logic to only the last case. Please correct me if this is wrong. Judging from the changes here, the issue should probably be called 'Allow fall though comments anywhere in switch blocks'.
> > 2. We must generate regression reports for any code changes to a check, please do this.
> > 3. I would expect a significant documentation update, since we are fundamentally changing the behavior of this check.
> 
> @nrmancuso Issue Description needs to be updated here is the more info for this issue: [#13553 (comment)](https://github.com/checkstyle/checkstyle/issues/13553#issuecomment-1752129573)

@romani Please verify this

**Comment by romani** [comment]
@nrmancuso , 

> The title of the issue is False positive in FallThroughCheck on last case, but I don't see anything limiting the new logic to only the last case

Issues are always created with knowledge and emotions on moment of creation. We do not know how it will end up in PR, so no updates before PR, during PR, maintainers should update issue (only maintainer has permissions to change issues created by others). If you see that implementation evolved in far bigger scope, please just update issue title or description.
only title is participating in release notes generation so to make it more explicit what is going to change update of title is what users might catch first. In most cases they just see name of Check in relese notes and open issue for details. So most critical part of title is name of module, as definition it was changed/affected.


>We must generate regression reports for any functional changes to a check, please do this.

@aayushRedHat , yes, please do.

>I would expect a significant documentation update, since we are fundamentally changing the behavior of this check.

Not significant, but it would be very good to update doc to let users read doc to understand how it works.
I recommend to finish functional review of this PR, come to pre-approval state and request doc update. To make sure we update doc on already ready and good functional fix.
@aayushRedHat , please generate report and think on doc update to explain (some words + example) behavior. 
@nrmancuso , after report is provided, please continue functional review.

**Comment by aayushRedHat** [comment]
> @nrmancuso ,
> 
> > The title of the issue is False positive in FallThroughCheck on last case, but I don't see anything limiting the new logic to only the last case
> 
> Issues are always created with knowledge and emotions on moment of creation. We do not know how it will end up in PR, so no updates before PR, during PR, maintainers should update issue (only maintainer has permissions to change issues created by others). If you see that implementation evolved in far bigger scope, please just update issue title or description. only title is participating in release notes generation so to make it more explicit what is going to change update of title is what users might catch first. In most cases they just see name of Check in relese notes and open issue for details. So most critical part of title is name of module, as definition it was changed/affected.
> 
> > We must generate regression reports for any functional changes to a check, please do this.
> 
> @aayushRedHat , yes, please do.
> 
> > I would expect a significant documentation update, since we are fundamentally changing the behavior of this check.
> 
> Not significant, but it would be very good to update doc to let users read doc to understand how it works. I recommend to finish functional review of this PR, come to pre-approval state and request doc update. To make sure we update doc on already ready and good functional fix. @aayushRedHat , please generate report and think on doc update to explain (some words + example) behavior. @nrmancuso , after report is provided, please continue functional review.

Sure @romani I will do by tomorrow

**Comment by nrmancuso** [comment]
@romani 

>Issues are always created with knowledge and emotions on moment of creation. We do not know how it will end up in PR, so no updates before PR, during PR, maintainers should ...

It seems like both you and @aayushRedHat misunderstood my concern at https://github.com/checkstyle/checkstyle/pull/14016#pullrequestreview-1738968086 (1), so let me rephrase it:

It is not clear from the changes to the code whether we are able to limit the change in check behavior to only the last case. If this PR does more than fix "False positive in FallThroughCheck on last case" I would like to understand what the goal is here. There is no point in functional review until a) the goal of the PR is clear, and b) until we have a diff report

**Comment by romani** [comment]
Github, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough9.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough9.java:21]: @nrmancuso,
we do have coverage for non last.

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough9.java:21]: > I would like to understand what the goal is here.

goal is to cover last case 100%, primary goal.
and looks like we able to cover not last too, we did not make very comprehensive coverage, but it is ok. This PR is already good step forward and we can create one more issue if we see some cases that are problematic to cover easily now.

**Comment by romani** [comment]
Appveyor failure is not related.
We need to bump to usage of 3.8.8 maven.
https://downloads.apache.org/maven/maven-3/

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
```
/home/semaphore/.m2/repository/R-4.27-202303020300/ is not found, downloading ... 02:20
--2023-11-26 21:28:02--  http://ftp-stud.fht-esslingen.de/pub/Mirrors/eclipse/eclipse/downloads/drops4/R-4.27-202303020300/ 02:20
Resolving ftp-stud.fht-esslingen.de (ftp-stud.fht-esslingen.de)... 129.143.116.10, 2001:7c0:700::10 02:20
Connecting to ftp-stud.fht-esslingen.de (ftp-stud.fht-esslingen.de)|129.143.116.10|:80... failed: Connection timed out. 04:30
Connecting to ftp-stud.fht-esslingen.de (ftp-stud.fht-esslingen.de)|2001:7c0:700::10|:80... failed: Network is unreachable.
```

it use to be green. 
folder exists - http://ftp-stud.fht-esslingen.de/pub/Mirrors/eclipse/eclipse/downloads/drops4/R-4.27-202303020300/

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
@nrmancuso , looks like we have problems with concurrent cancellation again.

@aayushRedHat , please generate report manually https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/README_MANUAL_EXECUTION.md#executing-diffgroovy while we investigating action problem

**Comment by nrmancuso** [comment]
Just sent a PR to remove all concurrency settings, this should unblock diff report generation: https://github.com/checkstyle/checkstyle/pull/14069

**Comment by aayushRedHat** [comment]
> Just sent a PR to remove all concurrency settings, this should unblock diff report generation: #14069

@nrmancuso Thanks, Please let me know if i need to manual create report after this

**Comment by romani** [comment]
Github, generate report

**Comment by nrmancuso** [comment]
GitHub, generate report

**Comment by romani** [comment]
Report generation is skipped again

**Comment by nrmancuso** [comment]
@aayushRedHat please investigate why this action is being skipped in your PRs

**Comment by aayushRedHat** [comment]
> @aayushRedHat please investigate why this action is being skipped in your PRs

Sure on it

**Comment by aayushRedHat** [comment]
GitHub, generate report

**Comment by nrmancuso** [comment]
@aayushRedHat please read through https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#report-generation and follow all directions there. You must provide a link to a raw text file for report generation.

**Comment by aayushRedHat** [comment]
Github, generate report

**Comment by nrmancuso** [comment]
@aayushRedHat you must have exactly one space between report parameter and url (no newlines, etc.)

**Comment by nrmancuso** [comment]
Github, generate report

**Comment by nrmancuso** [comment]
Github, generate report

**Comment by nrmancuso** [comment]
@aayushRedHat you have given a bad config, please make sure you are using https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/my_check.xml as a base

**Comment by aayushRedHat** [comment]
Github, generate report

**Comment by nrmancuso** [comment]
@aayushRedHat we have a few regressions, please create a new test/test input with cases found in regression and work through these.

**Comment by AayushSaini101** [comment]
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/d011b28_2023043149/reports/diff/index.html

@nrmancuso Can you help me to find the regression? I am new to read the regression, You can attached pics. I am happy to work on this:)

**Comment by nrmancuso** [comment]
@AayushSaini101 anywhere that you see a) violations were replaced by an exception (like [this](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/d011b28_2023043149/reports/diff/checkstyle/xref/home/runner/work/checkstyle/checkstyle/.ci-temp/contribution/checkstyle-tester/repositories/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/emptyblock/InputEmptyBlockCase.java.html#L1)) or a change in violations that result in a false positive or negative is a regression. Basically - any change in check behavior that is not expected. You will need to take time to read through the report and check differences.

**Comment by romani** [comment]
Please copy input code https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/d011b28_2023043149/reports/diff/sevntu-checkstyle/index.html#A2 to our test input s to not rely on regression testing too much and have good coverage just by running of our tests.

If you see other input code that is present in some project and not in our inputs, please copy.

**Comment by romani** [comment]
Diff report should have no exceptions, please keep adding cases that diff report found to our Inputs to detect exception by Junit execution.

**Comment by nrmancuso** [comment]
@aayushRedHat ping, let's get this across the finish line before we take up more issues. Example: https://github.com/checkstyle/checkstyle/pulls/aayushRedhat

**Comment by aayushRedHat** [comment]
> @aayushRedHat ping, let's get this across the finish line before we take up more issues. Example: https://github.com/checkstyle/checkstyle/pulls/aayushRedhat

Done Added test cases @romani @nrmancuso

**Comment by nrmancuso** [comment]
> Diff report should have no exceptions, please keep adding cases that diff report found to our Inputs to detect exception by Junit execution.

@aayushRedHat this is still not done. You must use code that is causing exceptions in the regression report, create test input that reproduces this behavior (exception), then fix your code such that we do not throw exceptions.

**Comment by aayushRedHat** [comment]
> > Diff report should have no exceptions, please keep adding cases that diff report found to our Inputs to detect exception by Junit execution.
> 
> @aayushRedHat this is still not done. You must use code that is causing exceptions in the regression report, create test input that reproduces this behavior (exception), then fix your code such that we do not throw exceptions.
> 
> I would expect a total of two (maybe three max) well named, targeted tests/test inputs in this PR. Naming like `BlahBlah10` is usually an indication of tests that are not well thought out.

@nrmancuso @romani I am facing an issue, for this test case: 
```
public class InputFallThrough20 {
   void method1(int a) {
        switch (a) {}
        int number = 0;
        switch (a) {
            case 1: number++; // violation 'Fall\ through from last branch of the switch statement.'
        }

        switch (a) {
            case 1:
        }

    }
}
```
when the check last group is true, i am getting 
```
Actual and expected violations differ.
expected to match: 15:(?:\d+:)?\s.*Fall\ through from last branch of the switch statement..*
but was          : 15:13: Fall through from the last branch of the switch statement.
```
I cannot able to get proper exp to path the position of the violation , what should I write  to compare properly.

 @Test
    public void test14() throws Exception {
        final String[] expected = {
            "15:" + getCheckMessage(MSG_FALL_THROUGH),
        };
        verifyWithInlineConfigParser(
                getPath("[InputFallThrough20.java](http://inputfallthrough20.java/)"),
                expected);
    }

**Comment by romani** [comment]
I do not see such failure in CI, please push code that you have problem with, to let us see whole context.

**Comment by aayushRedHat** [comment]
> I do not see such failure in CI, please push code that you have problem with, to let us see whole context.
Pushed @romani Please check InputFallThrough20 test file

**Comment by romani** [comment]
through from last branch
Vs
through from the last branch

Messages are slightly different

**Comment by aayushRedHat** [comment]
> through from last branch Vs through from the last branch
> 
> Messages are slightly different

Thanks @romani Tried to remove violation in reporti

**Comment by aayushRedHat** [comment]
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/a7f4e49_2023071036/reports/diff/index.html

@nrmancuso @romani Exceptions removed working on sonar fix

**Comment by aayushRedHat** [comment]
Github, generate report

**Comment by aayushRedHat** [comment]
@romani @nrmancuso Please review

**Comment by nrmancuso** [comment]
> @romani @nrmancuso Please review

Second point from https://github.com/checkstyle/checkstyle/pull/14016#issuecomment-1864514146 still stands, it is very unclear what the point of all these tests is. Each test should have some well defined purpose and reason to exist.

**Comment by aayushRedHat** [comment]
> > @romani @nrmancuso Please review
> 
> Second point from [#14016 (comment)](https://github.com/checkstyle/checkstyle/pull/14016#issuecomment-1864514146) still stands, it is very unclear what the point of all these tests is. Each test should have some well defined purpose and reason to exist.
Do i need to add for all test cases, There are lots of test cases i have added

**Comment by aayushRedHat** [comment]
> > @romani @nrmancuso Please review
> 
> Second point from [#14016 (comment)](https://github.com/checkstyle/checkstyle/pull/14016#issuecomment-1864514146) still stands, it is very unclear what the point of all these tests is. Each test should have some well defined purpose and reason to exist.

Done @nrmancuso

**Comment by aayushRedHat** [comment]
Github, generate report

**Comment by romani** [comment]
Github, generate report

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/Example1.java
Comment (romani) [inline src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/Example1.java:22]: We discourage usage of ok comments, if you want to keep it, put some wording after "ok" to explain why no violation is expected here.

Apply to all other cases.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughWithoutBreakStatements.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughWithoutBreakStatements.java:29]: I don't think all this code matters , please remove all code from input that doesn't matter

**Review by romani — CHANGES_REQUESTED** [review]
Minors:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughWithoutBreakStatements.java
Comment (aayushRedHat) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughWithoutBreakStatements.java]: Done @romani

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/Example1.java
Comment (aayushRedHat) [reply src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/Example1.java]: Done @romani

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java
Comment (nrmancuso) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java:365]: Please name this test and input in some meaningful way.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:417]: Please rename this, it is not obvious what this means.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:418]: What is the significance of this line?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:420]: `line == prevLineNumber` why does this condition matter?

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java:420]: Also, what types of DetailASTs are we checking in here? Only comments? Can we reuse `getNextNonCommentAst` within this loop?

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Tests are looking better, but still need some improvement.

Items:

**Comment by nrmancuso** [comment]
PR is abandoned, I am closing it.

**Pull request closed without merging**
